linked_list/linked_list.py

```diff
@@ -13,67 +13,125 @@ def __init__(self):
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        count = 0
+        current = self.head
+        while current != None:
+            count += 1
+            current = current.next
+        return count
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current = self.head
+        current_index = 0
+
+        while current != None:
+            if current_index == index:
+                return current.value
+            current_index += 1
+            current = current.next
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) or O(1)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+        current = self.head
+        while current.next != None:
+            current = current.next
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) or O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node =  Node(value)
+        if self.head == None:
+            self.head = new_node
+            return new_node
+        current = self.head
+        
+        while current.next != None:
+            current = current.next
+        current.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+        current = self.head
+        current_max = self.head.value
+        
+        
+        while current != None:
+            if current_max < current.value:
+                current_max = current.value
+            current = current.next
+        return current_max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        current = self.head
+        prev = None
+        while current != None:
+            if current.value == value:
+                if current == self.head:
+                    self.head = self.head.next
+                    current = self.head

```

Comment:
🤔 At this point, I would recommend exiting the function. Otherwise, you will continue iterating through and end up deleting _all_nodes found with the specified value. This method only asks you to delete the first.
```suggestion
                    return

```